pinot-tools/src/main/java/org/apache/pinot/tools/PinotDataAndQueryAnonymizer.java

```diff
@@ -0,0 +1,1287 @@
+      if (!_globalDictionaryColumns.containsKey(columnName)) {
+        // global dictionary will be built only for columns participating in filters
+        continue;
+      }
+      int totalCardinality = _globalDictionaryColumns.get(columnName);
+      ColumnMetadata columnMetadata = entry.getValue();
+      Dictionary dictionary = immutableSegment.getDictionary(columnName);
+      if (dictionary != null) {
+        // column has dictionary
+        for (int dictId = 0; dictId < columnMetadata.getCardinality(); dictId++) {
+          Object origValue = dictionary.get(dictId);
+          addOrigValueToGlobalDictionary(origValue, columnName, columnMetadata, totalCardinality);
+        }
+      } else {
+        // we build global dictionary only for columns that appear in filter
+        // and such columns should ideally always have a dictionary in segment
+        // so we should never end up here
+        throw new UnsupportedOperationException("Data generator currently does not support filter columns without dictionary");
+      }
+    }
+  }
+
+  /**
+   * First step towards building global dictionary
+   * by inserting the original values from segments
+   * into global dictionary
+   * @param origValue original value
+   * @param column column name
+   * @param columnMetadata column metadata
+   * @param cardinality total cardinality of column
+   */
+  private void addOrigValueToGlobalDictionary(
+      Object origValue,
+      String column,
+      ColumnMetadata columnMetadata,
+      int cardinality) {
+    _origToDerivedValueGlobalDictionary.putIfAbsent(column, new OrigAndDerivedValueHolder(columnMetadata.getDataType(), cardinality));
+    OrigAndDerivedValueHolder holder = _origToDerivedValueGlobalDictionary.get(column);
+    holder.addOrigValue(origValue);
+  }
+
+  /**
+   * This is the second step where we complete the global dictionaries
+   * by sorting the original values to get sort order across all segments
+   */
+  private void sortOriginalValuesInGlobalDictionaries() {
+    for (Map.Entry<String, OrigAndDerivedValueHolder> entry : _origToDerivedValueGlobalDictionary.entrySet()) {
+      OrigAndDerivedValueHolder valueHolder = entry.getValue();
+      valueHolder.sort();
+    }
+  }
+
+  /**
+   * This is the third and final step where we complete the global
+   * dictionaries by generating values:
+   *
+   * For numeric columns, we generate in order since
+   * we start with a base value.
+   * For string column, we first generate and then sort
+   */
+  private void addDerivedValuesToGlobalDictionaries() {
+    // update global dictionary for each column by adding
+    // the corresponding generated value for each orig value
+    for (Map.Entry<String, OrigAndDerivedValueHolder> entry : _origToDerivedValueGlobalDictionary.entrySet()) {
+      OrigAndDerivedValueHolder valueHolder = entry.getValue();
+      Object[] derivedValues = generateDerivedValuesForGlobalDictionary(valueHolder);
```

Comment:
can the generate method also set the derived values? It has the valueholder anyway

Reply:
done